### PR TITLE
Implemented query panel in latest posts block.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
Ths PR aims to implement the query panel specified in issue https://github.com/WordPress/gutenberg/issues/2662.

The panel is added to latest posts block. But this leaves a discussion point, it allows, for example, to show 8 oldest posts by setting sorting from oldest to newest.
To solve this we need to make a decision : 
We rename this block to "Posts" allowing to show the most recent ones or oldest or other option the user chooses.
Or we maintain the most recent posts concept, we always retrieve from the database the N most recent posts and then we sort this recent posts in accordance with the sorting criteria the user selected.
Feel free to share your options.

## How Has This Been Tested?
Try to change the values of the 3 different options in the query panel. Verify that the posts in the editor update in accordance with the option selected. Save the post on each option change, and open the post, verify the server side render also shows the selected option correctly.

## Screenshots (jpeg or gifs if applicable):
screen shot 2017-10-27 at 12 16 01
screen shot 2017-10-27 at 12 16 35
screen shot 2017-10-27 at 12 16 23

## Notes
There is a bug where component looks like loading forever if there are no posts that become more visible with this changes (because we can more easily select options without posts). A fix is being issued in https://github.com/WordPress/gutenberg/pull/3180.

The function buildTermsTree in utils was extracted from HierarchicalTermSelector component. As a future PR, this component will be refactored to make use of the utils function (removing the duplicate version) and will also be simplified/refactored to make use of the new component TermTreeSelect created in this PR.

### CODE REVIEW

File: blocks/term-tree-select/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/**
+ * External Dependencies
+ */
+import { unescape as unescapeString, repeat, reduce } from 'lodash';
+
+import SelectControl from '../inspector-controls/select-control';
+
+function getSelectOptions( terms, level = 0 ) {
+	return reduce( terms, ( options, term ) => (
```
Comment (aduth) [inline blocks/term-tree-select/index.js]: You might consider [Lodash's `flatMap`](https://lodash.com/docs/4.17.4#flatMap) which I recently found in #3185 to perform much better than `Array#reduce` + `Array#concat`.

Further, we can use array spread syntax to simplify this a bit:

```js
function getSelectOptions( terms, level = 0 ) {
	return flatMap( terms, ( term ) => [
		{
			value: term.id,
			label: repeat( '\u00A0', level * 3 ) + unescapeString( term.name ),
		},
		...getSelectOptions( term.children, level + 1 ),
	] );
}
```

File: blocks/term-tree-select/index.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+		options.concat( [ {
+			value: term.id,
+			label: repeat( '\u00A0', level * 3 ) + unescapeString( term.name ),
+		} ] ).concat( getSelectOptions( term.children, level + 1 ) )
+	), [] );
+}
+
+export default function TermTreeSelect( { termsTree, label, noOptionLabel, selectedTerm, onChange } ) {
+	const options = ( noOptionLabel ? [ { value: '', label: noOptionLabel } ] : [] ).concat( getSelectOptions( termsTree ) );
```
Comment (aduth) [inline blocks/term-tree-select/index.js]: This is a long line and difficult to read, I would suggest breaking it down a bit.

```js
const options = (
	noOptionLabel ? 
		[ { value: '', label: noOptionLabel } ] :
		[]
).concat( getSelectOptions( termsTree ) );
```

You could also use [Lodash's `compact`](https://lodash.com/docs/4.17.4#compact) to drop falsey values:

```js
const options = compact( [
	noOptionLabel && [ { value: '', label: noOptionLabel } ],
	...getSelectOptions( termsTree ),
] );
```

File: editor/utils/terms.js [hunk 3]
```diff
@@ -0,0 +1,27 @@
+/**
```
Comment (aduth) [inline editor/utils/terms.js]: Can we get some unit tests for this?

**Review by aduth — COMMENTED** [review]
Personally I'd have preferred to see the Latest Posts block updated to use [`withAPIData`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-api-data) before enhancing it to handle additional data attributes.

**Comment by jorgefilipecosta** [comment]
Thank you for your suggestions @aduth. I think I addressed all of them excluding an identification in php file. I made some changes to make things look better, but type item in categories is still not aligned with other types of other attributes, as the other type elements are not siblings (they are "cousins") if we add more spaces we break a phpcs rule.

File: editor/utils/terms.js (hunk 3 above)
Comment (jorgefilipecosta) [reply editor/utils/terms.js]: Added :) they were missing at the time because  I was not certain if a utils function was right for this case.

File: blocks/library/latest-posts/index.js [hunk 4]
```diff
@@ -72,22 +70,16 @@ registerBlockType( 'core/latest-posts', {
 		}
 
 		componentWillReceiveProps( nextProps ) {
-			const { postsToShow: postToShowCurrent } = this.props.attributes;
-			const { postsToShow: postToShowNext } = nextProps.attributes;
-			const { setAttributes } = this.props;
-
-			if ( postToShowCurrent === postToShowNext ) {
+			if ( this.props.attributes === nextProps.attributes ) {
```
Comment (aduth) [inline blocks/library/latest-posts/index.js]: I don't think we want to rely on a strict equality check here; these two objects may in fact be different references while containing the same values for `postsToShow`, `order`, `orderBy`, `categories`. If we need to, we can use [Lodash's `_.isEqual`](https://lodash.com/docs/4.17.4#isEqual) for a deep equality comparison.

File: blocks/query-panel/category-select.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+			onChange={ onChange }
+		/>
+	);
+}
+
+const applyWithAPIData = withAPIData( () => ( {
+	categories: '/wp/v2/categories',
+} ) );
+
+export default flowRight( [
```
Comment (aduth) [inline blocks/query-panel/category-select.js]: The `flowRight` here is unnecessary at this point (with a single higher-order component) and could be written as:

```js
export default applyWithAPIData( CategorySelect );
```

File: blocks/query-panel/index.js [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+	orderBy,
+	maxItems = DEFAULT_MAX_ITEMS,
+	minItems = DEFAULT_MIN_ITEMS,
+	onCategoryChange,
+	onNumberOfItemsChange,
+	onOrderChange,
+	onOrderByChange,
+} ) {
+	return (
+		<div>
```
Comment (aduth) [inline blocks/query-panel/index.js]: If the `div` is not serving a purpose on its own, wondering if we ought to just return as an array, with the only gotcha being we'd have to add keys to each rendered element, but this would reduce the total number of DOM nodes on the screen and reduce indentation in the component:

```js
return [
	<SelectControl
		key="..."
	// ...
];
```

**Comment by jorgefilipecosta** [comment]
Hi @aduth, thank you for your feedback 👍 , I rebased and addressed all the points.

File: blocks/library/latest-posts/index.php [hunk 7]
```diff
@@ -62,28 +65,48 @@ function gutenberg_render_block_core_latest_posts( $attributes ) {
...
-		'align'           => array(
-			'type'    => 'string',
-			'default' => 'center',
-		),
-	),
+register_block_type('core/latest-posts', array(
 	'render_callback' => 'gutenberg_render_block_core_latest_posts',
-) );
+	'attributes'      =>
+		array(
```
Comment (aduth) [inline blocks/library/latest-posts/index.php]: Why was this changed to put array opening on a line of its own? This isn't a style I've seen common to WordPress, and adds quite a bit of excessive indentation and awkward arrow spacing.

File: blocks/library/latest-posts/index.js [hunk 8]
```diff
@@ -72,22 +75,16 @@ registerBlockType( 'core/latest-posts', {
 		}
 
 		componentWillReceiveProps( nextProps ) {
-			const { postsToShow: postToShowCurrent } = this.props.attributes;
-			const { postsToShow: postToShowNext } = nextProps.attributes;
-			const { setAttributes } = this.props;
-
-			if ( postToShowCurrent === postToShowNext ) {
+			if ( isEqual( this.props.attributes, nextProps.attributes ) ) {
```
Comment (aduth) [inline blocks/library/latest-posts/index.js]: Just noting that we really want to check that a change in the attributes specific to the fetching of posts are the only ones we want to trigger a new request for; as implemented, a change to the block's alignment, columns, layout will trigger a new request.

File: blocks/query-panel/index.js [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+						label: __( 'Newest to Oldest' ),
+						value: 'date/desc',
+					},
+					{
+						label: __( 'Oldest to Newest' ),
+						value: 'date/asc',
+					},
+					{
+
+						/* translators: label for ordering posts by title in ascending order, translation may be required for languages using non latin alphabet */
```
Comment (aduth) [inline blocks/query-panel/index.js]: Kinda feel we can drop the rest of the comment after the first comma, that's really up to the translator to judge for themselves, and the first bit should be sufficient for understanding the intent of the string.

Comment (aduth) [reply blocks/query-panel/index.js]: Personal preference would be to drop the newline prior to this comment, feels in line with this ESLint rule we have enabled:

https://eslint.org/docs/rules/padded-blocks (never)

File: blocks/query-panel/index.js [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+		onCategoryChange && (
+			<CategorySelect
+				key="query-panel-category-select"
+				label={ __( 'Category' ) }
+				noOptionLabel={ __( 'All' ) }
+				selectedCategory={ category }
+				onChange={ onCategoryChange }
+			/> ),
+		onNumberOfItemsChange &&
+				(
```
Comment (aduth) [inline blocks/query-panel/index.js]: Inconsistency here with the previous condition, and also tabbed at least once more than necessary. Would suggest:

```jsx
		onNumberOfItemsChange && (
			<RangeControl
				key="query-panel-range-control"
				label={ __( 'Number of items' ) }
				value={ numberOfItems }
				onChange={ onNumberOfItemsChange }
				min={ minItems }
				max={ maxItems }
			/>
		),
```

**Review by aduth — CHANGES_REQUESTED** [review]
Requests incurred from unrelated attribute changes should be addressed.

File: blocks/library/latest-posts/index.php (hunk 7 above)
Comment (jorgefilipecosta) [reply blocks/library/latest-posts/index.php]: It was a try to make things look more aligned while respecting phpcs rules. I reverted this change but unfortunately, I'm not seeing a way to make things more aligned and respect phpcs.

**Review by karmatosed — CHANGES_REQUESTED** [review]
Design wise approving. I did find a bug though in changing the number of items. It got stuck after going to higher number then 1.

**Comment by jorgefilipecosta** [comment]
Hi @karmatosed, I was able to replicate something similar to what you described. When increasing number of posts to show the posts were not refreshing, leaving the user to be stuck in one post if a value of one was set before. It was introduced by a typo I made while addressing the last reviews and was fixed. Thank you for reporting the problem and for reviewing the design :)

File: editor/utils/terms.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+			return {
+				...term,
+				children: children && children.length ?
+					fillWithChildren( children ) :
+					[],
+			};
+		} );
+	};
+
+	const ret = fillWithChildren( termsByParent[ 0 ] || [] );
```
Comment (mcsf) [inline editor/utils/terms.js]: Would `termsByParent[ '0' ]` be more correct? Since it's dict key.

**Review by mcsf — COMMENTED** [review]
I need to run, so here's a more drive-by review.

**Comment by jorgefilipecosta** [comment]
Hi @mcsf, thank you a lot for your review every feedback was applied.

File: editor/utils/test/terms.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
+/**
+ * Internal dependencies
+ */
+import { buildTermsTree } from '../terms';
+
+describe( 'buildTermsTree()', () => {
+	describe( 'isHorizontalEdge', () => {
```
Comment (mcsf) [inline editor/utils/test/terms.js]: Some copy-paste that went wrong? 😄

Comment (jorgefilipecosta) [reply editor/utils/test/terms.js]: Yes @mcsf  I missed it, it was solved :D

File: blocks/query-panel/category-select.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ * WordPress dependencies
+ */
+import { withAPIData } from '@wordpress/components';
+
+/**
+ * Internal dependencies
+ */
+import { buildTermsTree } from '../../editor/utils/terms';
```
Comment (mcsf) [inline blocks/query-panel/category-select.js]: `buildTermsTree` is in `@wordpress/editor`, this should be a WordPress dependency.

Comment (jorgefilipecosta) [reply blocks/query-panel/category-select.js]: Moved to @wordpress/utils :)

Comment (gziolo) [reply blocks/query-panel/category-select.js]: Good catch @mcsf, blocks library shouldn't depend on the editor. 👍
